### PR TITLE
feat: odaiba grouped evaluation

### PR DESCRIPTION
**Pull request by Owen-Liuyuxuan**
## Summary

Adds **grouped-by-map-area closed-loop validation** for Odaiba Sakuraf, consuming portable `scenario_classification_json` from [tier4/Diffusion-Planner-Meta-Repository#25](https://github.com/tier4/Diffusion-Planner-Meta-Repository/pull/25).

### Key changes

- **`valid_predictor_closed_loop.py --mode grouped`** — one full-route rollout per bag, then per-episode metrics + videos by map area.
- **`scenario_generation/grouped_closed_loop_eval.py`** — reusable grouped runner (CLI + training).
- **Per-area metrics** — centerline deviation, turn-indicator match, safety clearance; `is_skipped` frames excluded from metrics only (not from video).
- **Training integration** — `closed_loop_validate()` in `train.py` auto-uses classification JSON when available; falls back to legacy full-route CL otherwise.
- **W&B logging** — per-area / per-metric-group scalars, results table, episode videos.

Depends on classification JSON produced by Meta-Repository `classify_scenario_corpus`.

---

### Step 1 — Label map areas (Meta-Repository)

```bash
source ~/pilot-auto.x2/install/setup.bash
cd ~/Diffusion-Planner-Meta-Repository/dataset/lanelet_area_tools
python labeler.py --map ~/autoware_map/.../lanelet2_map.osm
```

Save `lanelet2_map.*.areas.json`. Details: Meta-Repository PR #25.

### Step 2 — Scenario mapping + classification JSON (Meta-Repository)

```bash
# Edit dataset/configs/scenario_group_mapping.json if needed

cd ~/Diffusion-Planner-Meta-Repository
python -m dataset.classify_scenario_corpus \
  --labels ~/autoware_map/.../lanelet2_map.miraikan-teleport.areas.json \
  --npz_root /media/.../valid/2026-01-15 \
  --mapping dataset/configs/scenario_group_mapping.json \
  --dataset_name x2_dev/2231_odaiba_shinagawa_copied_from_xx1
```

Output: `dataset/scenario_classification_json/<dataset_name>/<date>.json`

### Step 3 — Offline closed-loop eval on `best_model.pth`

```bash
cd ~/Diffusion-Planner
source .venv/bin/activate   # CUDA torch + ffmpeg required

python diffusion_planner/valid_predictor_closed_loop.py \
  --mode grouped \
  --model_path ./best_model.pth \
  --npz_root /media/ukenryu/external_ssd/sakuraf_0623_full_sequence_x2_odaiba_valid/valid/2026-01-15 \
  --classification_json ../Diffusion-Planner-Meta-Repository/dataset/scenario_classification_json/x2_dev/2231_odaiba_shinagawa_copied_from_xx1/2026-01-15.json \
  --out_dir ./odaiba_cl_results \
  --near_miss_thresh 0.3 \
  --replan_interval 40 \
  --draw_every 8 \
  --device cuda
```

**`--mode full`** (default) keeps the original full-route segmented validation (no JSON required).

#### Output directory layout (`--out_dir ./odaiba_cl_results`)

```
odaiba_cl_results/
├── summary.json                 # run metadata + grouped_summary aggregates
├── metrics_summary.json         # per metric_group + per area_name aggregates
├── results_table.csv            # one row per episode (bag × area visit)
├── by_area/
│   ├── teleport_station_area_metrics_summary.json
│   ├── zeikan_curve_metrics_summary.json
│   └── ...
├── rollouts/                    # full-route rollout PNGs (one dir per bag)
│   ├── 13-42-45/
│   │   ├── 00000.png
│   │   └── ...
│   ├── 13-49-19/
│   └── ...
└── videos/                      # continuous-span episode MP4s
    ├── private_area/
    │   ├── teleport_station_area/
    │   │   ├── 13-42-45_0_<f_start>_<f_end>.mp4
    │   │   └── ...
    │   └── miraikan-entrance/
    ├── left_turn/
    │   ├── teleport-ekimae-intersection-left-turn/
    │   └── ...
    ├── curve/
    ├── straight/
    ├── right_turn/
    └── avoidance/
```

Each `results_table.csv` row includes: `metric_group`, `area_name`, `bag`, `span_index`, `segment`, `n_steps_run`, centerline / turn-indicator / collision metrics, `video_path`.

### Step 4 — Training / fine-tuning + W&B grouped validation *(not tested yet)*

On checkpoint save cadence (`save_utd`), rank-0 runs grouped CL when a classification JSON resolves:

```bash
python diffusion_planner/train_predictor.py \
  --exp_name odaiba_cl_grouped \
  --save_dir ./checkpoints/odaiba_cl_grouped \
  --train_set_list /path/to/train.txt \
  --valid_set_list /path/to/valid.txt \
  --use_wandb true \
  --closed_loop_npz_root /media/.../valid/2026-01-15 \
  --closed_loop_scenario_dataset_name x2_dev/2231_odaiba_shinagawa_copied_from_xx1 \
  --closed_loop_near_miss_thresh 0.3 \
  --closed_loop_grouped_wandb_max_videos 24
```

**JSON resolution (graceful fallback):**

1. `--closed_loop_classification_json /path/to/2026-01-15.json` if file exists
2. Auto: `$SCENARIO_CLASSIFICATION_JSON_ROOT` or `../Diffusion-Planner-Meta-Repository/dataset/scenario_classification_json/<dataset_name>/<npz_date>.json`
3. Else → legacy **full-route** closed-loop (unchanged behavior; warning printed)

**On-disk (per checkpoint):** `epochXXXX/closed_loop/grouped/` (same structure as Step 3).

**W&B keys (grouped mode):**

| Key | Content |
|-----|---------|
| `closed_loop/mode` | `"grouped"` or `"full"` |
| `closed_loop/grouped/n_episodes`, `n_bags`, `elapsed_sec` | Run summary |
| `closed_loop/grouped/metric_group/<group>/<metric>` | Aggregates per scenario type |
| `closed_loop/grouped/area/<area>/<metric>` | Aggregates per map area |
| `closed_loop/grouped/results_table` | `wandb.Table` of episode rows |
| `closed_loop/grouped/video/<...>` | Episode MP4s (capped by `--closed_loop_grouped_wandb_max_videos`, default 24) |

Full-route fallback keeps legacy keys: `closed_loop/collision_segment_rate`, `closed_loop/video/<stem>`, etc.

---

## Architecture

```
Meta-Repository                          Diffusion-Planner
───────────────                          ─────────────────
labeler.py → *.areas.json
scenario_group_mapping.json
classify_scenario_corpus.py
    → scenario_classification_json/
           └── 2026-01-15.json  ──────►  valid_predictor_closed_loop.py --mode grouped
                                         train.py closed_loop_validate() → wandb
```

Eval flow per bag:

1. One **full-route** closed-loop rollout (`rollouts/<bag>/`).
2. Tag each sim step by reproducer frame index → area (from JSON `labeled_ranges`).
3. One metrics row + one video per **episode** (`video_start_idx`…`video_end_idx`).
4. Metrics skip `is_skipped` labeled frames; videos include the full continuous span.

---

## Test plan

- [x] `python diffusion_planner/valid_predictor_closed_loop.py --mode grouped` on Odaiba `2026-01-15` + Meta-Repo JSON
- [x] Verify continuous episode videos include stopping / unlabeled gaps
- [x] Unit tests: `scenario_generation/tests/test_scenario_classification.py`
- [x] `--mode full` smoke via `diffusion_planner/valid_predictor_closed_loop.py` (single bag + `--seg_len 150`)
- [ ] Training run with `--closed_loop_npz_root` + `--closed_loop_scenario_dataset_name`; confirm W&B grouped metrics
- [ ] Training run **without** classification JSON; confirm full-route fallback + warning

### CODE REVIEW

**Comment by Owen-Liuyuxuan** [comment]
```
timing (slowest first):
  draw                      487.164s    4844 calls   100.57 ms/call
  rollout                   346.548s       4 calls  86636.90 ms/call
  pre_step                  130.555s   38734 calls     3.37 ms/call
  timeline_load_npz          44.587s   14464 calls     3.08 ms/call
  score_safety               42.255s   38726 calls     1.09 ms/call
  model_forward              28.299s     988 calls    28.64 ms/call
  build_videos               19.502s      37 calls   527.09 ms/call
  cursor_step                 2.370s   38726 calls     0.06 ms/call
  timeline_load_poses         1.546s       4 calls   386.54 ms/call
  advance                     1.023s   38726 calls     0.03 ms/call
  replan_plan_cache           0.968s   37738 calls     0.03 ms/call
  to_torch                    0.956s     988 calls     0.97 ms/call
  aggregate_metrics           0.348s      37 calls     9.40 ms/call
  model_forward_cached        0.000s   37738 calls     0.00 ms/call
  TOTAL (summed)           1106.120s  over 57101 steps
grouped closed-loop @epoch 81: 13911 steps, coll=0, near_miss=1 in 368.1s

```

File: diffusion_planner/diffusion_planner/utils/train_utils.py [hunk 1]
```diff
@@ -71,17 +72,42 @@ def get_epoch_mean_loss(epoch_loss):
     return epoch_mean_loss
 
 
+def strip_module_prefix(state_dict: dict[str, Any]) -> dict[str, Any]:
```
Comment (Owen-Liuyuxuan) [inline diffusion_planner/diffusion_planner/utils/train_utils.py:85]: This is not good, we should preserve the original setting. and force us to only launch with ddp (or add a gate to handle this special case when we launch in single thread)

File: diffusion_planner/diffusion_planner/utils/train_utils.py [hunk 2]
```diff
@@ -113,7 +139,7 @@ def resume_model(path: str, model, optimizer, scheduler, ema, device):
         wandb_id = None
 
     try:
-        ema.ema.load_state_dict(ckpt["ema_state_dict"])
+        ema.ema.load_state_dict(strip_module_prefix(ckpt["ema_state_dict"]))
```
Comment (Owen-Liuyuxuan) [inline diffusion_planner/diffusion_planner/utils/train_utils.py]: This is also the case and we should assume ddp mainly

File: diffusion_planner/diffusion_planner/train_epoch.py [hunk 3]
```diff
@@ -33,6 +33,10 @@ def heading_to_cos_sin(x):
 
 
 def train_epoch(data_loader, model, optimizer, args, ema, aug: StatePerturbation = None):
+    if len(data_loader) == 0:
```
Comment (Owen-Liuyuxuan) [inline diffusion_planner/diffusion_planner/train_epoch.py:36]: Keep this for testing until we finish the PR. (After reviewing the PR, this guard will be deleted)

File: diffusion_planner/diffusion_planner/validate_model.py [hunk 4]
```diff
@@ -278,6 +278,27 @@ def validate_model(model, val_loader, args, return_pred=False) -> tuple[float, f
     """return: ave_loss_ego, ave_loss_neighbor"""
     device = args.device
     model.eval()
+
+    if len(val_loader) == 0:
```
Comment (Owen-Liuyuxuan) [inline diffusion_planner/diffusion_planner/validate_model.py:282]: Keep this for testing until we finish the PR

File: diffusion_planner/diffusion_planner/utils/ddp.py [hunk 5]
```diff
@@ -16,8 +16,8 @@ def ddp_setup_universal(verbose=False, args=None):
         rank = int(os.environ["RANK"])
         world_size = int(os.environ["WORLD_SIZE"])
         gpu = int(os.environ["LOCAL_RANK"])
-        os.environ["MASTER_PORT"] = str(getattr(args, "port", "29529"))
-        os.environ["MASTER_ADDR"] = "localhost"
+        os.environ.setdefault("MASTER_ADDR", "127.0.0.1")
```
Comment (Owen-Liuyuxuan) [inline diffusion_planner/diffusion_planner/utils/ddp.py]: I don't know, my account on sakuraf work perfectly with the default env

**Comment by Owen-Liuyuxuan** [comment]
multi GPU worked

image


But GPU is not the bottleneck, the CPU drawing and roll out is a harder constraints. But anyway we can go multi-threading.

**Comment by Owen-Liuyuxuan** [comment]
## How closed-loop eval is distributed under DDP

### When does multi-GPU CL run?

At the end of each checkpoint-save epoch (`save_utd`), if `--closed_loop_npz_root` is set **and** training is launched with `torchrun` + `--ddp true` (`world_size > 1`), **every rank** calls `closed_loop_validate()` with the same already-loaded DDP model.

- No extra checkpoint reload
- No `ProcessPoolExecutor` / per-GPU subprocess pool
- Single-GPU (`--ddp false` or `NPROC_PER_NODE=1`): only rank 0 runs CL

```python
# train.py (simplified)
if ddp_cl:
    torch.distributed.barrier()
    closed_loop_validate(diffusion_planner, args, epoch, cl_out)  # all ranks
elif global_rank == 0:
    closed_loop_validate(...)
```

Output path uses `args.save_dir` (not rank-0-only `save_path`) so all ranks can write shard files:
`{save_dir}/epochXXXX/closed_loop/grouped/...`

---

### What unit is sharded?

Sharding is at **bag** level inside **grouped** closed-loop eval (classification JSON `time_series` + NPZ bags under `closed_loop_npz_root`).

1. `closed_loop_validate()` discovers eval jobs from NPZ root + classification JSON.
2. For each job → `run_grouped_closed_loop_eval(..., ddp_rank=rank, ddp_world_size=world_size)`.
3. Bags are sorted, then split round-robin:

```python
# grouped_closed_loop_ddp.py
def shard_bag_items(bag_items, rank, world_size):
    return [bag_items[i] for i in range(rank, len(bag_items), world_size)]
```

**Example — 4 bags, 4 GPUs** (`2026-01-15` smoke):

| Rank | GPU (`LOCAL_RANK`) | Bags |
|------|-------------------|------|
| 0 | cuda:0 | bag[0] |
| 1 | cuda:1 | bag[1] |
| 2 | cuda:2 | bag[2] |
| 3 | cuda:3 | bag[3] |

**Example — 4 bags, 2 GPUs:**

| Rank | Bags |
|------|------|
| 0 | bag[0], bag[2] |
| 1 | bag[1], bag[3] |

Each assigned bag still runs a **full route rollout** (all NPZ frames in that bag, subject to `replan_interval` / `draw_every`). There is **no** frame-level sharding across GPUs.

Per rank:
- **Device**: `cuda:{LOCAL_RANK}` via `ddp_device_for_rank()`
- **Model**: `ddp.get_model(model, args.ddp)` → inner `.module` for inference
- **Partial output**: `out_dir/ddp_shards/rank_{rank:03d}.json` (+ videos for that rank’s bags under `grouped/`)

---

### Merge on rank 0

After all ranks finish:

```python
dist.barrier()
if rank == 0:
    merged = merge_ddp_grouped_shards(job_out, world_size, ...)
    summary = merge_grouped_eval_summaries(merged_jobs)
    wandb.log(...)
```

`merge_ddp_grouped_shards()` reads every `rank_*.json`, concatenates segments / video paths / step counts / per-bag timing, then writes the usual grouped outputs (`summary.json`, `results_table.csv`, `timing.json`, …). Only rank 0 logs W&B and prints the final `(DDP xN)` summary.

---

### 8 GPUs, 4 bags (idle ranks)

Round-robin with `world_size=8` and 4 bags:

| Rank | Bags assigned |
|------|----------------|
| 0 | bag[0] |
| 1 | bag[1] |
| 2 | bag[2] |
| 3 | bag[3] |
| 4–7 | **none** (empty shard) |

Ranks 4–7:
- `shard_bag_items` returns `[]` → rollout loop is a no-op
- Still call `write_ddp_shard(...)` with empty `segments` / `video_mp4s` / `total_steps=0`
- Still enter `dist.barrier()` with ranks 0–3

So idle GPUs do not hang: they finish quickly, write an empty shard file, and synchronize at the barrier like the busy ranks. Rank 0 merge skips missing shard files (`if not path.is_file(): continue`) but normally all 8 ranks write `rank_000.json` … `rank_007.json`.

Wall-clock time is dominated by the slowest bag among ranks 0–3, not by the idle ranks.

---

### What is *not* DDP-sharded

| Case | Behavior |
|------|----------|
| No classification JSON match | Legacy `run_closed_loop_eval()` — rank 0 only; other ranks return early when `ddp_active` |
| Train / valid dataloaders | Standard `DistributedSampler` — separate from CL |
| Frames inside one bag | Sequential on one GPU |
| Checkpoint save | Rank 0 only |

---

### Launch

```bash
# single GPU (smoke default)
bash diffusion_planner/run_closed_loop_wandb_smoke.sh

# multi-GPU bag sharding
NPROC_PER_NODE=4 bash diffusion_planner/run_closed_loop_wandb_smoke.sh
```

Requires **`torchrun` + `--ddp true` together**. `torchrun` with `--ddp false` spawns independent rank-0 jobs (duplicate W&B, all on GPU 0) — not this path.

---

### Checkpoint loading note (`resume_model`)

Most checkpoints are saved from DDP and keep the `module.` prefix:

- **`use_ddp=True`**: original tier4-main load (`model.load_state_dict(ckpt["model"])`)
- **`use_ddp=False`** (single-GPU smoke): `strip_module_prefix` before load

File: diffusion_planner/diffusion_planner/utils/train_utils.py (hunk 2 above)
Comment (Owen-Liuyuxuan) [reply diffusion_planner/diffusion_planner/utils/train_utils.py]: Solve

File: diffusion_planner/diffusion_planner/validate_model.py (hunk 4 above)
Comment (Owen-Liuyuxuan) [reply diffusion_planner/diffusion_planner/validate_model.py:282]: This is actually a good safeguard

File: diffusion_planner/diffusion_planner/train_epoch.py (hunk 3 above)
Comment (Owen-Liuyuxuan) [reply diffusion_planner/diffusion_planner/train_epoch.py:36]: This is actually a good safeguard

**Comment by Owen-Liuyuxuan** [comment]
I can use the "test/odaiba_group_eval" branch to run multi-gpu update

**Comment by Owen-Liuyuxuan** [comment]
@SakodaShintaro  This branch will be kept as my testing branch and demonstration of the my full pipeline. But I will try to divide this PR into smaller pieces so it will become easier for people to review.

This branch turn into draft now.

**Comment by Owen-Liuyuxuan** [comment]
Split into a 6-PR stack for deeper review (local worktrees + `.cursor/pr-split/` review docs):

| # | PR | Scope |
|---|-----|--------|
| 1/6 | https://github.com/tier4/Diffusion-Planner/pull/230 | Non-DDP checkpoint `module.` prefix strip |
| 2/6 | https://github.com/tier4/Diffusion-Planner/pull/231 | Empty train/valid dataloader guards |
| 3/6 | https://github.com/tier4/Diffusion-Planner/pull/232 | Shared `_closed_loop_replan_step` |
| 4/6 | https://github.com/tier4/Diffusion-Planner/pull/233 | `ClosedLoopEvaluation` class framework |
| 5/6 | https://github.com/tier4/Diffusion-Planner/pull/234 | `GroupedByAreaClosedLoopEvaluation` + suite |
| 6/6 | https://github.com/tier4/Diffusion-Planner/pull/235 | Training `closed_loop_validate` integration |

Merge order: **230 → 231 → 232 → 233 → 234 → 235**. This monolithic PR can close once the stack lands.

Note: branches are based on `tier4-main` @ `81f2983`; `origin/tier4-main` has since advanced to `8290101` — stack may need rebase before final merge.
